### PR TITLE
TreeView: fix setting of role when items are empty && add axe tests

### PR DESCRIPTION
**Pull request by marker-dao**

### CODE REVIEW

File: packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+fixture`TreeView: Common tests with axe`
+  .page(url(__dirname, '../../container.html'));
+
+const TREEVIEW_SELECTOR = '#container';
+
+[[], employees].forEach((items) => {
+  [true, false].forEach((searchEnabled) => {
+    ['none', 'normal', 'selectAll'].forEach((showCheckBoxesMode) => {
+      [null, 'no data text'].forEach((noDataText) => {
```
Comment (ksercs) [inline packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts]: We'll create a special util for a11y testing later in a separate PR.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.search.js [hunk 2]
```diff
@@ -86,6 +86,22 @@ const TreeViewSearch = TreeViewBase.inherit(searchBoxMixin).inherit({
         return this._itemContainer(this.option('searchEnabled'));
     },
 
+    _setAriaRoleToElement() {
+        const { items } = this.option();
+
+        if(items?.length) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.search.js]: Why should we do this **only** when search is enabled?

File: packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          searchEnabled,
+          showCheckBoxesMode,
+          noDataText,
+          items,
+          displayExpr: 'fullName',
+        }));
+      });
+    });
+  });
+});
```
Comment (ksercs) [inline packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts:38]: Now we have a new report when scrollable is rendered. I'll give more details in a personal conversation

File: packages/devextreme/js/ui/tree_view/ui.tree_view.search.js [hunk 4]
```diff
@@ -86,6 +86,22 @@ const TreeViewSearch = TreeViewBase.inherit(searchBoxMixin).inherit({
...
+            this.setAria({ role: 'tree' });
+        }
+    },
+
+    _renderFocusTarget() {
+        const { items, tabIndex } = this.option();
+
+        if(items?.length) {
+            this._focusTarget().attr({ tabindex: tabIndex });
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.search.js]: Why should we do this? "No text" is a part of a component, I believe we should provide an ability to focus this element to make it be able for screen reading.
AFAIR other components work this way, we should keep a consistent behavior.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.search.js (hunk 2 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/tree_view/ui.tree_view.search.js]: I refactored this part of code.

File: packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts (hunk 3 above)
Comment (marker-dao) [reply packages/devextreme/testing/testcafe/tests/navigation/treeView/accessibility.ts:38]: The report was fixed.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.search.js (hunk 4 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/tree_view/ui.tree_view.search.js]: I removed this part of code.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 5]
```diff
@@ -577,11 +586,41 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
+        const isContentHeightExceeded = scrollableContentHeight > scrollableContainerHeight;
+
+        if(!isContentHeightExceeded) {
+            return;
+        }
+
+        const { tabIndex } = this.option();
+        const $nodeContainer = this.$element().find(`.${NODE_CONTAINER_CLASS}`).first();
+
+        $nodeContainer.attr({ tabindex: tabIndex });
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: It's really strange for me to see that adding of a tabIndex depends on a calculated height of a nested scrollable content.
I believe there should be another more simple and common solution.

**Review by ksercs — CHANGES_REQUESTED** [review]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.scrolling.tests.js [hunk 6]
```diff
@@ -226,6 +229,30 @@ QUnit.module('useNativeScrolling', () => {
...
+
+        const isIOS = devices.real().ios;
+
+        assert.strictEqual($nodeContainer.attr('tabindex'), isIOS ? '0' : undefined);
+
+        wrapper.instance.option({ height: 100 });
+
+        $nodeContainer = $treeView.find(`.${NODE_CONTAINER_CLASS}`).first();
+
+        assert.strictEqual($nodeContainer.attr('tabindex'), isIOS ? '0' : undefined);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.scrolling.tests.js]: I don't see any IOS specificity affecting this task.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js (hunk 5 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: We decided to remove tabindex setting in this case.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.scrolling.tests.js (hunk 6 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.scrolling.tests.js]: These tests was removed.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 7]
```diff
@@ -577,11 +577,21 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
         this._renderScrollableContainer();
         this._renderEmptyMessage(this._dataAdapter.getRootNodes());
+
         this.callBase();
-        this.setAria('role', 'tree');
+
+        this._setAriaRoleToElement();
+    },
+
+    _setAriaRoleToElement() {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: I think that suffix is excess. It's clear enough that we add a role to an element.
```suggestion
    _setAriaRole() {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.markup.tests.js [hunk 8]
```diff
@@ -55,13 +55,17 @@ QUnit.module('aria accessibility', {
...
+        assert.strictEqual(this.$element.attr('role'), 'tree', 'role is correct');
+
+        this.instance.option({ items: [] });
+
+        assert.strictEqual(this.$element.attr('role'), undefined, 'role is not set');
     });
 
-    QUnit.test('scrollable should have role treeitem attribute', function(assert) {
-        const $scrollable = this.$element.find('.' + SCROLLABLE_CLASS);
+    QUnit.test('scrollable should not have role attribute', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.markup.tests.js]: I believe this test is not necessary now.
Seems strange to me that we test that element does not have an attribute.

Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.markup.tests.js]: I think this test could be useful if we want to set a role attribute in the future. Right now we know it's wrong. But we have added Axe tests, so I agree with you that we can get rid of this test.

Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.markup.tests.js]: [3fe0ac7ff44946354ba94acc5981fa062abc353d]

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js (hunk 7 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [17d788a031862755f5b9987fbe73de76e65e459c]

**Review by ksercs — APPROVED** [review]

**Review by Zedwag — APPROVED** [review]

**Pull request merged**